### ISSUE TITLE
LP-vault ledger counts a backing refill as both new principal and a recovery, over-stating available principal

### ISSUE THREAD

**Issue by 0x-SquidSol** [dcccrypto/percolator-prog#413]
## Summary

A deposit into a partially-consumed backing bucket is counted twice by the LP-vault ledger: once as new principal by the depositing handler, and again as a "recovery" the next time the ledger is synced. Reported available principal then exceeds the backing that actually exists, and redemption pays out against the inflated figure.

## Why the engine's counter cannot be read as a loss watermark

`consumed_liened_backing_num` is a *net outstanding* counter, not a lifetime-loss total. Consumption raises it alongside the provider's receivable (`src/v16.rs:1159-1178` in the engine):

```rust
bucket.consumed_liened_backing_num += amount;
source.spent_backing_num           += amount;
source.provider_receivable_num     += amount;
```

and a later deposit pays that receivable back down (`prepare_counterparty_backing_add_delta`, engine `src/v16.rs:1009-1018`; wrapper twin `add_fresh_counterparty_backing_view`, `src/v16_program.rs:16899-16906`):

```rust
let refill = amount.min(source.provider_receivable_num);
bucket.consumed_liened_backing_num -= refill;
source.provider_receivable_num     -= refill;
bucket.fresh_unliened_backing_num  += amount;   // the FULL amount, not amount - refill
```

The engine holds `provider_receivable_num == consumed_liened_backing_num` as an invariant (checked on entry at `src/v16.rs:996`), so the two move together by construction. A deposit therefore both clears part of the consumed counter **and** adds its full value to fresh backing. That behaviour is intended and is covered by your own test, `v16_wrapper_backing_top_up_refills_provider_receivable_in_engine` (`tests/v16_wrapper.rs:8098`) — we are not questioning it.

The defect is that the wrapper's ledger reads a decrease in that counter as evidence of recovery (`sync_backing_domain_ledger`, `src/v16_program.rs:9934-9946`):

```rust
let unavailable_atoms = backing_unavailable_principal_atoms(bucket)?;   // (consumed + impaired) / BOUND_SCALE
if unavailable_atoms >= ledger.last_observed_unavailable_principal_atoms {
    ledger.cumulative_loss_atoms += unavailable_atoms - last_observed;
} else {
    ledger.cumulative_recovery_atoms += last_observed - unavailable_atoms;
}
```

Only a refill ever lowers that total. So the `else` branch fires precisely when new money was paid in — never when backing was genuinely returned — while the depositing handler has separately credited the same atoms as principal (`handle_top_up_backing_bucket:10324`, `handle_deposit_to_lp_vault:14810`, both `total_principal_atoms += amount`).

Note this is not a broken invariant in `lp_vault_nav_atoms`. Its stated contract, `net impairment = loss - recovery == current unavailable principal` (`src/v16.rs:16488`), still holds. The break is upstream of it: `unavailable` under-counts lifetime consumption by the refill, while `total_principal` counts the deposit in full, so `available = principal - net_impairment` over-states by exactly the refill.

## Ordering

Both deposit handlers sync *before* depositing, so the drop is not observed on that instruction — the watermark is written at the pre-deposit value and the refill happens after:

```
handle_top_up_backing_bucket:  sync (:10315) → deposit (:10321) → principal += amount (:10324)
handle_deposit_to_lp_vault:    sync (:14797) → deposit (:14804) → principal += amount (:14810)
```

The spurious credit lands on the **next** sync. That is not a narrow window: both read paths re-sync against the live bucket before computing anything — `lp_vault_domain_nav_atoms:9996` and `lp_vault_domain_available_principal_atoms:10025` — so the inflation is applied by the very next NAV or redemption query, and is persisted by the next deposit's own pre-sync.

## Reproduction

Added to the existing `processor::tests` module, driving the real `sync_backing_domain_ledger` and the real `percolator::lp_vault::lp_vault_nav_atoms`. Passes against `main` at `19d5d93`:

```
cargo test --lib backing_refill_is_counted
test processor::tests::backing_refill_is_counted_as_both_principal_and_recovery ... ok
test result: ok. 1 passed; 0 failed
```

| step | bucket | ledger |
|---|---|---|
| provider deposits 40 | fresh 40, consumed 0 | principal 40 |
| engine consumes 40 for a trader's source-attributed gain | fresh 0, consumed 40 | `cumulative_loss = 40` (correct) |
| **a further 10 is deposited** | fresh 10, consumed 30 | principal 50 **and** `cumulative_recovery = 10` |

Final state: `lp_vault_nav_atoms` reports **20** atoms of available principal while the bucket holds **10**. Lifetime consumption was 40, but the ledger's net impairment is only 30 — the 10-atom gap is exactly the refill.

```rust
#[test]
fn backing_refill_is_counted_as_both_principal_and_recovery() {
    const SCALE: u128 = percolator::BOUND_SCALE;

    let mut ledger = state::BackingDomainLedgerAccountV16 {
        market_group: [7u8; 32],
        authority: [8u8; 32],
        domain: 2,
        ..Default::default()
    };
    let mut bucket = percolator::BackingBucketV16 {
        fresh_unliened_backing_num: 40 * SCALE,
        ..percolator::BackingBucketV16::EMPTY
    };

    // 1. Provider deposits 40. The caller books principal; sync sees nothing
    //    unavailable yet.
    ledger.total_principal_atoms = 40;
    sync_backing_domain_ledger(&mut ledger, &bucket).unwrap();
    assert_eq!(ledger.cumulative_loss_atoms, 0);
    assert_eq!(ledger.cumulative_recovery_atoms, 0);

    // 2. The engine consumes all 40 to cover a trader's source-attributed
    //    gain. Unavailable principal rises, so the ledger books a real loss.
    bucket.fresh_unliened_backing_num = 0;
    bucket.consumed_liened_backing_num = 40 * SCALE;
    sync_backing_domain_ledger(&mut ledger, &bucket).unwrap();
    assert_eq!(ledger.cumulative_loss_atoms, 40, "40 genuinely became unavailable");

    // 3. A further deposit of 10 arrives. The engine refills the provider's
    //    receivable first, so `consumed` drops by 10 while the full 10 also
    //    lands in fresh backing (see the engine's own
    //    `v16_wrapper_backing_top_up_refills_provider_receivable_in_engine`).
    bucket.consumed_liened_backing_num = 30 * SCALE;
    bucket.fresh_unliened_backing_num = 10 * SCALE;
    ledger.total_principal_atoms += 10; // the caller's own principal credit
    sync_backing_domain_ledger(&mut ledger, &bucket).unwrap();

    // The fall in unavailable principal is read as a recovery, even though
    // nothing was returned -- the depositor simply paid it in.
    assert_eq!(ledger.cumulative_recovery_atoms, 10, "refill mistaken for recovery");

    // Net effect: the ledger reports twice the backing that actually exists.
    let reported_available = percolator::lp_vault::lp_vault_nav_atoms(
        ledger.total_principal_atoms,
        ledger.total_earnings_atoms,
        ledger.total_earnings_withdrawn_atoms,
        ledger.cumulative_loss_atoms,
        ledger.cumulative_recovery_atoms,
        0,
    )
    .unwrap();
    let real_backing = bucket.fresh_unliened_backing_num / SCALE;

    assert_eq!(real_backing, 10, "only 10 atoms of backing are actually present");
    assert_eq!(reported_available, 20, "the ledger claims 20");
    assert_eq!(
        reported_available - real_backing,
        10,
        "10 atoms of phantom principal, exactly the refilled amount"
    );
}
```

The bucket transitions are stepped directly so the arithmetic is isolated and readable; each mirrors what the engine does on the corresponding instruction, and the refill step matches the counter movement in `prepare_counterparty_backing_add_delta` and the test cited above.

### The same numbers already appear in your end-to-end suite

We want to flag this prominently rather than bury it, because it cuts both ways.

`v16_wrapper_backing_domain_ledger_tracks_unavailable_principal_loss_and_recovery` (`tests/v16_wrapper.rs:7791-7882`) drives this exact sequence through real instructions — `TopUpBackingBucket { amount: 40 }`, `ConvertReleasedPnl { amount: 40 }`, `SyncBackingDomainLedger`, `TopUpBackingBucket { amount: 10 }`, `SyncBackingDomainLedger` — and asserts:

```rust
assert_eq!(ledger_state.total_principal_atoms, 50);
assert_eq!(ledger_state.cumulative_loss_atoms, 40);
assert_eq!(ledger_state.cumulative_recovery_atoms, 10);
assert_eq!(ledger_state.last_observed_unavailable_principal_atoms, 30);
```

Those are precisely the values our probe produces, which we think is the strongest available confirmation that the state is reachable through the real instruction path and that our stepped counters are faithful — 50 atoms were transferred in, 40 were paid out to the trader, 10 remain.

It also means this behaviour is currently asserted as correct, so a fix necessarily changes that assertion. **If the recovery credit on a top-up is intended, please treat this as a report about its consequence for `lp_vault_nav_atoms` rather than about the counter itself, and we will close it.** Our reason for filing anyway is that the test verifies the counter's movement without asserting anything about available principal, and a top-up is the *only* thing that lowers `consumed_liened_backing_num` — so as written, the `else` branch does not have a legitimate trigger. The other path that lowers unavailable principal, the terminal impaired drain (engine `src/v16.rs:1109-1124`), is documented as explicitly *not* re-crediting backing ("expiry already removed that backing once; re-crediting would double-count the senior backing"), so it is not a value recovery either.

## Impact

`handle_execute_redemption` pays `principal_portion = floor(shares * available_principal / total_shares)` (`src/v16_program.rs:15395`, `:15418`), with `available_principal` derived from the same inflated ledger. Redeemers are therefore paid pro-rata against principal that does not exist.

To be explicit about one alternative reading we checked and ruled out: the "two pots" the redemption comment refers to (`:15392-15393`) are the two **domains** of the asset — `lp_vault_combined_available_principal_atoms` sums the own and sibling domain (`:10042-10062`). There is no idle, undeployed pot that could legitimately account for the gap; the full deposit goes into the bucket.

The payout is not silently over-drawn from thin air — `principal_portion` becomes `backing_num` (`:15439`) and is withdrawn from fresh backing, which fails closed once that is exhausted (engine `src/v16.rs:1034-1039`). The consequence is therefore a redistribution followed by a revert, in exit order. Taking the state above with 50 shares outstanding:

- a holder of 25 shares redeems first: `principal_out = 25 * 20 / 50 = 10`, and the bucket holds exactly 10, so it succeeds and drains it. Their true share of real backing was 5.
- the remaining 25 shares are then unredeemable: `principal_out` is still 10, `fresh_unliened_backing_num` is 0, and the withdraw reverts.

So the first mover is paid roughly double their entitlement out of the second's, and the last holder's principal is stranded until someone tops the bucket up again — which, being a top-up, widens the gap further. The over-statement is cumulative: every refill after a consumption adds its own phantom amount.

The user-facing deposit path makes this worse than a bookkeeping drift. `handle_deposit_to_lp_vault` prices shares in Phase 1, off the pre-deposit NAV, and only then does the deposit inflate NAV by `amount + refill`. The phantom `refill` accrues to *all* shareholders rather than the depositor who funded it, so an ordinary deposit into a vault with outstanding consumed backing silently transfers value to incumbents and enlarges the shortfall that the last redeemers will hit. No privileged role is needed for that path, and nothing has to be misconfigured — only the ordinary sequence of backing being consumed (a trader realising source-attributed positive PnL) and then replenished.

We rate this high: a value-accounting break reachable through normal operation, with LP principal as the loss surface.

## Suggested direction

The sync needs to distinguish "backing returned by the engine" from "backing newly paid in", which the bucket total alone cannot express. Two shapes, and we would defer to you on which fits:

- **Have the depositing handlers tell the ledger.** They already know `amount` and can compute `refill = min(amount, provider_receivable_num)` before the deposit, then advance `last_observed_unavailable_principal_atoms` down by that refill so the next sync sees no drop. A genuine engine-side return would still register.
- **Sync against the receivable directly.** `source.provider_receivable_num` is what the refill actually pays down, and the engine already pins it equal to `consumed_liened_backing_num`. Deriving recovery from a change in the receivable *net of deposits* separates the two causes at the source.

Whichever shape you prefer, note that all of them change the assertion in `v16_wrapper_backing_domain_ledger_tracks_unavailable_principal_loss_and_recovery`, so that test needs a deliberate decision rather than a mechanical update. We would rather you make that call than have us presume it, which is why we have filed this without an accompanying pull request.

Our own preference is the third option below, because it is the only one that also covers the terminal impaired drain, and because it makes `cumulative_loss_atoms` genuinely monotone — which is what the farm-snapshot semantics at `:9929-9933` appear to want.

A narrower alternative: make the loss/recovery branch one-sided like the earnings branch directly above it (`:9920-9927`), so a fall in unavailable principal only updates the watermark without crediting recovery. That under-reports genuine recoveries, which is the conservative direction, but it would need checking against the farm-snapshot semantics documented at `:9929-9933` — that comment already states recoveries are tracked separately and must not decrement `cumulative_loss_atoms`, so the two counters' roles are load-bearing elsewhere.

## Environment

- `dcccrypto/percolator-prog` `main` at `19d5d93`
- Engine `dcccrypto/percolator` `main` at `b5ddba2`
- `cargo test --lib`, rustc 1.97.1. Wrapper lib suite green with the probe added (22 passing).

**Comment by 0x-SquidSol** [comment]
Independent verification against the **deployed baseline** (wrapper `15eb8b0c`, engine `1bea6692`). TL;DR: the over-statement is **real**, but I traced the redemption payout end-to-end and it is **clamped** — so this is an accounting-correctness bug, not an over-withdrawal / fund-loss path.

### The over-statement is confirmed real
The refill is double-counted. `handle_deposit_to_lp_vault` (and `handle_top_up_backing_bucket`) bump `total_principal` by the refill `R`, and the next `sync_backing_domain_ledger` also bumps `cumulative_recovery_atoms` by `R` for the `consumed_liened`/`unavailable` reduction the same refill caused — `v16_program.rs:10041-10043` (recovery branch). So `available = total_principal − (loss − recovery)` counts `R` twice. It is momentarily correct right after the deposit (sync runs pre-mutation) and only over-states at the *next* sync.

### But it is not exploitable into an over-withdrawal
The inflated `available` feeds only the **principal** term of a redemption (it cancels out of the earnings term, since `earnings = atoms − principal` and both move with `available`). `handle_execute_redemption` then gates the actual SPL release on **physical** engine/vault state, not on the ledger:

- `bucket.fresh_unliened_backing_num < backing_num` → reject (`v16_program.rs:15787`), where `backing_num = principal_portion · BOUND_SCALE`. The bucket only ever holds physically deposited/released backing, so cumulative principal paid out ≤ physical `fresh_unliened`. The extra `R` can never leave the vault.
- post-redemption `credit_rate_num != CREDIT_RATE_SCALE` → reject (`v16_program.rs:15915`), so no outstanding winner claim is left under-backed by a redemption.
- `validate_shape` → `validate_header_aggregate_totals` re-checks `senior_with_backing ≤ vault` (`v16.rs:5564-5590`).

So the counter is wrong but never releases unbacked value. The only residual is an intra-vault first-mover redistribution when `positive_claim_bound_num == 0` (winner already settled/withdrawn), and even then it is bounded by physical backing (`:15787`) — a later LP's redemption reverts rather than any net drain.

### Suggestion
Unless there's a redemption path that reads `available`/NAV **without** the `fresh_unliened_backing_num` clamp, this looks like it should be **accounting-correctness**, not fund-loss — if the current framing implies over-withdrawal, that reads as over-claimed. Fix is local: don't credit `cumulative_recovery` for a `consumed_liened` reduction that was caused by a provider *refill* (net the refilled amount out of the recovery delta in the same instruction that bumped `total_principal`), at `v16_program.rs:10041-10043`.

Happy to be corrected if I'm missing a path.

_(Found during an independent two-pass security review; verified against deployed bytes, not `main`.)_
